File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.py
from collections import Counter
from typing import List

class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        # Step 1: Count frequencies
        freq = Counter(nums)
        
        # Step 2: Create buckets (index = frequency)
        buckets = [[] for _ in range(len(nums) + 1)]
        for num, count in freq.items():
            buckets[count].append(num)
        
        # Step 3: Collect top k from highest frequency
        ans = []
        for i in range(len(buckets) - 1, 0, -1):
            for num in buckets[i]:
                ans.append(num)
                if len(ans) == k:
                    return ans

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.cpp
#include <vector>
#include <unordered_map>
using namespace std;

class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        // Step 1: Count frequencies
        unordered_map<int, int> freq;
        for (int num : nums) {
            freq[num]++;
        }

        // Step 2: Create buckets (index = frequency)
        vector<vector<int>> buckets(nums.size() + 1);
        for (auto& pair : freq) {
            int num = pair.first;
            int count = pair.second;
            buckets[count].push_back(num);
        }

        // Step 3: Collect top k from highest frequency
        vector<int> ans;
        for (int i = buckets.size() - 1; i > 0; i--) {
            for (int num : buckets[i]) {
                ans.push_back(num);
                if (ans.size() == k) {
                    return ans;
                }
            }
        }

        return ans; // fallback (though normally returned earlier)
    }
};